dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class DeprecationWarningChannel implements Channel {
+    private static final Logger log = LoggerFactory.getLogger(DeprecationWarningChannel.class);
+    private static final Object SENTINEL = new Object();
+
+    private final Channel delegate;
+    private final TaggedMetricRegistry registry;
+    private final Cache<String, Object> loggingRateLimiter = Caffeine.newBuilder()
```

Comment:
why not use the Guava rate limiter?

Reply:
I need/want a RateLimiter that's namespaced to a particular serviceName, Guava RateLimiters are limiting per instance. One option would be to have the cache be String -> RateLimiter, but Caffeine is perhaps sufficient. Maybe I don't need to be so cautious with memory use/object allocation?

Reply:
probably fine either way